### PR TITLE
feat: wrench model predictive controller on 2D spacecraft

### PR DESCRIPTION
**Pull request by joris997**
To complete a nice set of initial controllers, I've started on an MPC that publishes thrusts and torques (the wrench, separated over two publishers as per the PX4 message types).
Currently, the robot spins around its z-axis so I still need to fix that.

### CODE REVIEW

File: px4_mpc/px4_mpc/mpc_spacecraft.py [hunk 1]
```diff
@@ -190,20 +216,32 @@ def publish_reference(self, pub, reference):
...
-        thrust_command = thrust_rates  # NOTE: Tune in thrust multiplier
+        thrust_command = thrust_rates[0:3] * 0.07  # NOTE: Tune in thrust multiplier
         rates_setpoint_msg = VehicleRatesSetpoint()
         rates_setpoint_msg.timestamp = int(Clock().now().nanoseconds / 1000)
-        rates_setpoint_msg.roll = float(thrust_rates[3])
-        rates_setpoint_msg.pitch = -float(thrust_rates[4])
-        rates_setpoint_msg.yaw = -float(thrust_rates[5])
+        rates_setpoint_msg.roll = 0.0
+        rates_setpoint_msg.pitch = 0.0
+        rates_setpoint_msg.yaw = 0.0
```
Comment (Jaeyoung-Lim) [inline px4_mpc/px4_mpc/mpc_spacecraft.py:226]: We don't want to disable the rate publishing?

**Review by Jaeyoung-Lim — CHANGES_REQUESTED** [review]

File: px4_mpc/px4_mpc/mpc_spacecraft.py [hunk 2]
```diff
@@ -81,8 +87,8 @@ def __init__(self):
             depth=1
         )
 
-        # Get mode; rate, body force/torque, direct_allocation
-        self.mode = 'direct_allocation'
+        # Get mode; rate, wrench, direct_allocation
+        self.mode = 'wrench'
```
Comment (Jaeyoung-Lim) [inline px4_mpc/px4_mpc/mpc_spacecraft.py:91]: Can you make this a parameter of the launchfile?

**Review by Jaeyoung-Lim — CHANGES_REQUESTED** [review]

File: px4_mpc/px4_mpc/mpc_spacecraft.py [hunk 3]
```diff
@@ -241,13 +279,18 @@ def cmdloop_callback(self):
         offboard_msg.position = False
         offboard_msg.velocity = False
         offboard_msg.acceleration = False
+        offboard_msg.position = False
+        offboard_msg.velocity = False
+        offboard_msg.acceleration = False
```
Comment (Jaeyoung-Lim) [inline px4_mpc/px4_mpc/mpc_spacecraft.py:284]: Redundant?

File: px4_mpc/px4_mpc/controllers/spacecraft_wrench_mpc.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
+############################################################################
+#
+#   Copyright (C) 2022 PX4 Development Team. All rights reserved.
```
Comment (Jaeyoung-Lim) [inline px4_mpc/px4_mpc/controllers/spacecraft_wrench_mpc.py:3]:
```suggestion
#   Copyright (C) 2024 PX4 Development Team. All rights reserved.
```

**Comment by Pedro-Roque** [comment]
@joris997 you might need to merge the latest master to your branch to make sure it is up to date.

File: px4_mpc/px4_mpc/mpc_spacecraft.py (hunk 1 above)
Comment (joris997) [reply px4_mpc/px4_mpc/mpc_spacecraft.py]: the rate publishing is never called upon. both rate and allocation functions remain but in the `cmdloop_callback` it is checked which publishers align to the controller mode.
The change here was actually due to a merge conflict that slipped through. It's now set to the original rates.

**Comment by Jaeyoung-Lim** [comment]
> @joris997 you might need to merge the latest master to your branch to make sure it is up to date.

I would suggest a rebase

**Review by Jaeyoung-Lim — APPROVED** [review]
LGTM, thanks!

**Comment by Jaeyoung-Lim** [comment]
@joris997 Could you add instructions on how to configure the different modes of the MPC?

**Pull request merged**